merkle/smt/layout.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smt
+
+import "sort"
+
+// Layout defines the mapping between tree nodes and tiles.
+//
+// The tree is divided into horizontal tiers of the heights specified in the
+// constructor. The tiers are counted down from the tree root. Nodes located at
+// the tier borders are roots of the corresponding tiles. A tile includes all
+// the nodes strictly below the tile root, within a single tier.
+type Layout struct {
+	depths []int
+}
+
+// NewLayout creates a tree layout based on the passed-in tier heights.
+func NewLayout(heights []int) Layout {
+	depths := make([]int, len(heights)+1)
+	for i, h := range heights {
+		if h <= 0 {
+			panic("tile heights must be positive")
+		}
```

Comment:
We use `uint` below, so I'd just suck this up and make the inputs to this `uint` too so that it's properly compile-time checked and thus avoid this panic.

Reply:
Yeah, I did this first, but note that we should still check != 0. Also, there is no `SearchInts` for uints, so I kinda chose int to avoid another cast. But that is an implementation details, and from the "API" point of view `uint` indeed makes more sense - fixed.